src/tasks/tasks.hpp

```diff
@@ -0,0 +1,493 @@
+          }
+          return TaskStatus::complete;
+        },
+        exec_limits));
+    last_task = tasks.back().get();
+  }
+
+  template <class... Args>
+  TaskID AddTask(TaskID dep, Args &&...args) {
+    return AddTask(TaskQualifier::normal, dep, std::forward<Args>(args)...);
+  }
+
+  template <class... Args>
+  TaskID AddTask(const TaskQualifier tq, TaskID dep, Args &&...args) {
+    assert(tq.Valid());
+
+    // user-space tasks always depend on something. if no dependencies are given,
+    // make the task dependent on the list's first_task
+    if (dep.empty()) dep = TaskID(first_task);
+
+    if (!tq.Once() || (tq.Once() && unique_id == 0)) {
+      AddUserTask(dep, std::forward<Args>(args)...);
+    } else {
+      tasks.push_back(std::make_shared<Task>(
+          dep, [=]() { return TaskStatus::complete; }, exec_limits));
+    }
+
+    Task *my_task = tasks.back().get();
+    TaskID id(my_task);
+
+    if (tq.LocalSync() || tq.GlobalSync() || tq.Once()) {
+      regional_tasks.push_back(my_task);
+    }
+
+    if (tq.GlobalSync()) {
+      bool do_mpi = false;
+#ifdef MPI_PARALLEL
+      // make status, request, and comm for this global task
+      global_status.push_back(std::make_shared<int>(0));
+      global_request.push_back(std::make_shared<MPI_Request>(MPI_REQUEST_NULL));
+      // be careful about the custom deleter so it doesn't call
+      // an MPI function after Finalize
+      global_comm.emplace_back(new MPI_Comm, [&](MPI_Comm *d) {
+        int finalized;
+        MPI_Finalized(&finalized);
+        if (!finalized) MPI_Comm_free(d);
+      });
+      // we need another communicator to support multiple in flight non-blocking
+      // collectives where we can't guarantee calling order across ranks
+      MPI_Comm_dup(MPI_COMM_WORLD, global_comm.back().get());
+      do_mpi = true;
+#endif // MPI_PARALLEL
+      TaskID start;
+      // only call MPI once per region, on the list with unique_id = 0
+      if (unique_id == 0 && do_mpi) {
+#ifdef MPI_PARALLEL
+        // add a task that starts the Iallreduce on the task statuses
+        tasks.push_back(std::make_shared<Task>(
+            id,
+            [my_task, &stat = *global_status.back(), &req = *global_request.back(),
+             &comm = *global_comm.back()]() {
+              // jump through a couple hoops to figure out statuses of all instances of
+              // my_task accross all lists in the enclosing TaskRegion
+              auto dependent = my_task->GetDependent(TaskStatus::complete);
+              assert(dependent.size() == 1);
+              auto mytask = *dependent.begin();
+              stat = 0;
+              for (auto dep : mytask->GetDependencies()) {
+                stat = std::max(stat, static_cast<int>(dep->GetStatus()));
+              }
+              MPI_Iallreduce(MPI_IN_PLACE, &stat, 1, MPI_INT, MPI_MAX, comm, &req);
+              return TaskStatus::complete;
+            },
+            exec_limits));
```

Comment:
Is there a reason this can't be written using `parthenon::AllReduce`? It might simplify the code a little bit, but maybe not really worth making a change.

Reply:
I'm not sure.  I'd have to think about that.  I suspect it's possible, but there might be some gotcha that's not immediately obvious.